rust/src/nasl/builtin/network/socket.rs

```diff
@@ -379,62 +205,80 @@ impl NaslSockets {
+                    Some(timeout) => {
+                        if timeout < 1 {
+                            conn.read(&mut data)?
+                        } else {
+                            conn.read_with_timeout(&mut data, Duration::from_secs(timeout as u64))?
+                        }
+                    }
+                    None => conn.read(&mut data)?,
+                };
 
-                let mut result = conn.socket.recv_from(data.as_mut_slice());
+                Ok(NaslValue::Data(data[..pos].to_vec()))
+            }
+            NaslSocket::Close => Err(FunctionErrorKind::WrongArgument(
+                "the given socket FD is already closed".to_string(),
+            )),
+        }
+    }
 
-                for _ in 0..NUM_TIMES_TO_RESEND {
-                    match result {
-                        Ok((size, origin)) => {
-                            if conn.socket.peer_addr()? == origin {
-                                data.truncate(size);
-                                ret = Ok(NaslValue::Data(data));
-                                break;
-                            }
+    #[nasl_function(named(socket, length, timeout))]
+    fn recv_line(
+        &self,
+        socket: usize,
+        #[allow(unused_variables)] length: usize,
+        timeout: Option<i64>,
+    ) -> Result<NaslValue, FunctionErrorKind> {
+        let mut data = String::new();
+        match self
+            .handles
+            .write()
+            .unwrap()
+            .handles
+            .get_mut(socket)
+            .ok_or(FunctionErrorKind::WrongArgument(format!(
+                "the given socket FD {socket} does not exist"
+            )))? {
+            NaslSocket::Tcp(conn) => {
+                let pos = match timeout {
+                    Some(timeout) => {
+                        if timeout < 1 {
+                            conn.read_line(&mut data)?
+                        } else {
+                            conn.read_line_with_timeout(
+                                &mut data,
+                                Duration::from_secs(timeout as u64),
+                            )?
                         }
-                        Err(e) => match e.kind() {
-                            io::ErrorKind::TimedOut => {
-                                conn.socket.send(&conn.buffer)?;
-                            }
-                            kind => {
-                                ret = Err(FunctionErrorKind::IOError(kind));
-                                break;
-                            }
-                        },
-                    };
-
-                    result = conn.socket.recv_from(data.as_mut_slice());
-                }
-                if let Some(timeout) = old {
-                    conn.socket.set_read_timeout(Some(timeout))?;
-                }
-                ret
+                    }
+                    None => conn.read_line(&mut data)?,
+                };
```

Comment:
Same structure as above, although it's `read_line` here, so I think this is fine but I would replace it with something like
```rust
let pos = match convert_timeout(timeout) { 
    Some(timeout) => conn.read_with_timeout(&mut data[pos..], timeout), 
    None => conn.read(&mut data[pos..]), 
}?;
```